cupyx/scipy/sparse/coo.py

```diff
@@ -95,9 +95,10 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
         elif _scipy_available and scipy.sparse.issparse(arg1):
             # Convert scipy.sparse to cupyx.scipy.sparse
             x = arg1.tocoo()
-            data = cupy.array(x.data)
-            row = cupy.array(x.row, dtype='i')
-            col = cupy.array(x.col, dtype='i')
+            import cupy.array_api as cpx
+            data = cpx.asarray(x.data)
```

Comment:
There's a problem now that `cupyx.sparse` can't operate the sparse matrices correctly if data and indices are not of type `cupy.ndarray`.

```py
In [1]: import cupyx.scipy.sparse

In [2]: import cupy

In [3]: import scipy.sparse

In [4]: a = scipy.sparse.rand(100, 100, density=0.5)
/home/yashchuk/.conda/envs/scipy-meson/lib/python3.10/site-packages/scipy/sparse/_sputils.py:94: UserWarning: The numpy.array_api submodule is still experimental. See NEP 47.
  import cupy.array_api as cpx

In [5]: type(a)
Out[5]: scipy.sparse._coo.coo_matrix

In [6]: type(a.data)
Out[6]: numpy.ndarray

In [7]: type(a.row)
Out[7]: numpy.ndarray

In [8]: cupyx.scipy.sparse.coo_matrix(a)
Out[8]: <cupyx.scipy.sparse.coo.coo_matrix at 0x7fa7482e6230>

In [9]: a_cp = cupyx.scipy.sparse.coo_matrix(a)

In [10]: a_cp.T
Out[10]: <cupyx.scipy.sparse.coo.coo_matrix at 0x7fa746ded540>

In [11]: a_cp + a_cp
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [11], in <module>
----> 1 a_cp + a_cp

File ~/amd-project/cupy/cupyx/scipy/sparse/base.py:97, in spmatrix.__add__(self, other)
     96 def __add__(self, other):
---> 97     return self.tocsr().__add__(other)

File ~/amd-project/cupy/cupyx/scipy/sparse/coo.py:541, in coo_matrix.tocsr(self, copy)
    538 # copy is silently ignored (in line with SciPy) because both
    539 # sum_duplicates and coosort change the underlying data
    540 x = self.copy()
--> 541 x.sum_duplicates()
    542 cusparse.coosort(x, 'r')
    543 x = cusparse.coo2csr(x)

File ~/amd-project/cupy/cupyx/scipy/sparse/coo.py:413, in coo_matrix.sum_duplicates(self)
    407     return
    408 # Note: The sorting order below follows the cuSPARSE convention (first
    409 # row then col, so-called row-major) and differs from that of SciPy, as
    410 # the cuSPARSE functions such as cusparseSpMV() assume this sorting
    411 # order.
    412 # See https://docs.nvidia.com/cuda/cusparse/index.html#coo-format
--> 413 keys = cupy.stack([self.col, self.row])
    414 order = cupy.lexsort(keys)
    415 from cupy.array_api.array_compatibility import get_namespace

File ~/amd-project/cupy/cupy/_manipulation/join.py:135, in stack(tup, axis, out)
    122 def stack(tup, axis=0, out=None):
    123     """Stacks arrays along a new axis.
    124
    125     Args:
   (...)
    133     .. seealso:: :func:`numpy.stack`
    134     """
--> 135     return concatenate([cupy.expand_dims(x, axis) for x in tup], axis, out)

File ~/amd-project/cupy/cupy/_manipulation/join.py:135, in <listcomp>(.0)
    122 def stack(tup, axis=0, out=None):
    123     """Stacks arrays along a new axis.
    124
    125     Args:
   (...)
    133     .. seealso:: :func:`numpy.stack`
    134     """
--> 135     return concatenate([cupy.expand_dims(x, axis) for x in tup], axis, out)

File ~/amd-project/cupy/cupy/_manipulation/dims.py:148, in expand_dims(a, axis)
    146 if type(axis) not in (tuple, list):
    147     axis = axis,
--> 148 return _manipulation._expand_dims(a, axis)

TypeError: Argument 'a' has incorrect type (expected cupy._core.core.ndarray, got Array)
```

Reply:
I think member variables of sparse matrices (data, indices) should remain normal CuPy arrays in CuPy and normal NumPy arrays in SciPy.
When we construct the sparse matrices in SciPy we should dispatch based on the type of arrays passed into the constructor:
```
import scipy.sparse
import numpy
import cupy
import numpy.array_api as npx
import cupy.array_api as cpx

# NumPy inputs
row, col = numpy.array([0]), numpy.array([1])
data = numpy.array([0.5])
a_np = scipy.sparse.coo_matrix((data, (row, col))) # normal scipy matrix with numpy data

# CuPy inputs
row, col = cupy.array([0]), cupy.array([1])
data = cupy.array([0.5])
a_cp = scipy.sparse.coo_matrix((data, (row, col))) # raises an error, we might not need to cover this case now

# NumPy Array API inputs
row, col = npx.asarray([0]), npx.asarray([1])
data = npx.asarray([0.5])
a_npx = scipy.sparse.coo_matrix((data, (row, col))) # raises an error, we probably need to get this working
```

In the future I see `scipy.sparse.coo_matrix` and other constructors in the sparse module to be functions and use a dispatch layer based on uarray/singledispatch/etc. At the moment to get the prototype working we can have inside SciPy something like:
```py
def scipy.sparse.coo_matrix(arg1, shape=None, dtype=None, copy=False):
    # check that __array_namespace__ exists,
    # let's dispatch only for cupy.array_api arrays now, not for normal cupy arrays
    namespace = get_namespace(data, row, col)
    if "cupy.array_api" == namespace.__name__:
        return cupyx.scipy.coo_matrix((data._array, (row._array, col._array)))
    elif "numpy.array_api" == namespace.__name__:
        return normal_scipy_coo_matrix((data._array, (row._array, col._array)))
    else:
        return normal_scipy_coo_matrix((data, (row, col)))
```